### PR TITLE
Formula Validation using ExpressionParser

### PR DESCRIPTION
**Pull request by rishi-aga**
One of the PR that Resolves https://github.com/yahoo/elide/discussions/1829

## Description
Formula Validation using ExpressionParser.

## License
I confirm that this contribution is made under an Apache 2.0 license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/queryengines/sql/expression/ExpressionParser.java [hunk 1]
```diff
@@ -50,6 +88,17 @@ public ExpressionParser(MetaDataStore store) {
...
 
         for (String referenceName : referenceNames) {
+
+            // Change `sql from='joinName' column='columnName[a1:v1][a2:v2]'` to `joinName.columnName`
+            if (referenceName.startsWith(SQL_HELPER_PREFIX)) {
+                try {
+                    Template template = handlebars.compileInline(toFormulaReference(referenceName));
+                    referenceName = template.apply(Collections.emptyMap());
+                } catch (IOException e) {
+                    // Do Nothing
```
Comment (aklish) [inline elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/queryengines/sql/expression/ExpressionParser.java:98]: Maybe throw an IllegalStateException here?

File: elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/queryengines/sql/metadata/SQLReferenceTable.java [hunk 2]
```diff
@@ -69,13 +69,16 @@ protected SQLReferenceTable(MetaDataStore metaDataStore, Set<Queryable> queryabl
...
-        queryables.stream().forEach(queryable -> {
-            Queryable next = queryable;
-            do {
-                resolveAndStoreAllReferencesAndJoins(next);
-                next = next.getSource();
-            } while (next.isNested());
-        });
+        queryables
+           .stream()
+           .filter(queryable -> queryable.isNested() || queryable.isRoot())
```
Comment (aklish) [inline elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/queryengines/sql/metadata/SQLReferenceTable.java:74]: What's this filter for?

File: elide-datastore/elide-datastore-aggregation/src/test/java/com/yahoo/elide/datastores/aggregation/custom/DailyAverageScorePerPeriod.java [hunk 3]
```diff
@@ -56,7 +56,7 @@ public QueryPlan resolve(Query query) {
                 .metricProjection(SQLMetricProjection.builder()
                         .alias(getAlias())
                         .name(getName())
-                        .expression(getExpression())
+                        .expression("AVG({{$highScore}})")
```
Comment (aklish) [inline elide-datastore/elide-datastore-aggregation/src/test/java/com/yahoo/elide/datastores/aggregation/custom/DailyAverageScorePerPeriod.java:59]: Isn't this the same as before?  Why the change?

**Review by aklish — CHANGES_REQUESTED** [review]

File: elide-datastore/elide-datastore-aggregation/src/test/java/com/yahoo/elide/datastores/aggregation/custom/DailyAverageScorePerPeriod.java (hunk 3 above)
Comment (rishi-aga) [reply elide-datastore/elide-datastore-aggregation/src/test/java/com/yahoo/elide/datastores/aggregation/custom/DailyAverageScorePerPeriod.java:59]: added `$`

File: elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/queryengines/sql/metadata/SQLReferenceTable.java (hunk 2 above)
Comment (rishi-aga) [reply elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/queryengines/sql/metadata/SQLReferenceTable.java:74]: Wanted to exclude adding those queryable for Dynamic tables which are already there in Static Table, Static Table's queryable already get the preference.
Expression Parser was failing for the scenario: When it is a Query over SQLTable, then Query's ColumnProjection may reference a logical column which is defined in SQLTable but not in Query. It fails at: https://github.com/yahoo/elide/blob/master/elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/queryengines/sql/expression/ExpressionParser.java#L81

File: elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/queryengines/sql/query/SQLColumnProjection.java [hunk 4]
```diff
@@ -41,6 +41,14 @@ default String toSQL(Queryable source, SQLReferenceTable lookupTable) {
     default boolean canNest(Queryable source, SQLReferenceTable lookupTable) {
         SQLDialect dialect = source.getConnectionDetails().getDialect();
         String sql = toSQL(source.getSource(), lookupTable);
+
+        // TODO: Remove this when switching to TableContext
+        // Resolved Reference would be empty if value is not provided in @MetricFormula as value is optional.
+        // Once we change this to use TableContext, then it should be able to resolve expression directly.
+        if (this instanceof SQLMetricProjection && sql.equals("")) {
```
Comment (aklish) [inline elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/queryengines/sql/query/SQLColumnProjection.java:48]: Instead of this hack here, why not just change DailyAverageScorePerPeriod to override canNest to true.

**Review by aklish — APPROVED** [review]
